### PR TITLE
test all

### PR DESCRIPTION
**Pull request by Lawliet-Chan**
Sync with @mask-pp and refer to https://github.com/ethereum/go-ethereum/blob/master/Makefile#L32 
Open tests for all modules.

### CODE REVIEW

**Comment by ChuhanJin** [comment]
Can one of the admins verify this patch?

**Comment by Lawliet-Chan** [comment]
@HAOYUatHZ @mask-pp  
Shall we need to remove `signer/core/signed_data_internal_test.go`? Seems there is no in https://github.com/ethereum/go-ethereum/tree/master/signer/core
BTW, when I open all module tests, there are many errors in CI. Are these errors normal?

**Comment by mask-pp** [comment]
> 

Fix those fails as many as possible.

**Comment by Lawliet-Chan** [comment]
There are many errors in [CI](https://github.com/scroll-tech/go-ethereum/actions/runs/3762501927/jobs/6395260373) as follows:
```
########## BAD BLOCK #########
Chain config: {ChainID: 19763 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: 0 Byzantium: 0 Constantinople: <nil> Petersburg: <nil> Istanbul: <nil>, Muir Glacier: <nil>, Berlin: <nil>, London: <nil>, Arrow Glacier: <nil>, Engine: ethash}

Number: 31
Hash: 0xe2f9dfee463001d2f78292b0fa631e9f89c9cc91fe747febccb4311687ff765d
	 0: cumulative: 88619 gas: 88619 contract: 0x880EC53Af800b5Cd051531672EF4fc4De233bD5d status: 1 tx: 0x447c5a664f0907bdc93ed211e0882adfcfa0943db9001baf407091d4d8286985 logs: [] bloom: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 state: 


Error: invalid merkle root (remote: ffa8aa7510636622cda6eb544d09ae5b251790c774f51a082303e874c9e83b9a local: fca8[613](https://github.com/scroll-tech/go-ethereum/actions/runs/3762501927/jobs/6395260373#step:5:614)67a36a81916c7eb864b34a3cf07a8d7d51112f6718462bed32360eb6c)
##############################
 
--- FAIL: TestEthSuite (0.80s)
    suite_test.go:40: could not run geth: invalid merkle root (remote: ffa8aa7510636622cda6eb544d09ae5b251790c774f51a082303e874c9e83b9a local: fca861367a36a81916c7eb864b34a3cf07a8d7d51112f6718462bed32360eb6c)
FAIL
FAIL	github.com/scroll-tech/go-ethereum/cmd/devp2p/internal/ethtest	0.908s
``` 

```
--- FAIL: TestT8n (0.25s)
    t8n_test.go:220: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/1/alloc.json --input.txs ./testdata/1/txs.json --input.env ./testdata/1/env.json --state.fork Frontier+1346
    test_cmd.go:281: (stderr:1) ERROR(3): failed constructing chain configuration: syntax error, invalid eip number 1346
    t8n_test.go:220: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/1/alloc.json --input.txs ./testdata/1/txs.json --input.env ./testdata/1/env.json --state.fork Byzantium
    test_cmd.go:281: (stderr:2) INFO [12-23|01:55:48.979] rejected tx                              index=1 hash=0557ba..18d673 from=0x8A8eAFb1cf62BfBeb1741769DAE1a9dd47996192 error="nonce too low: address 0x8A8eAFb1cf62BfBeb1741769DAE1a9dd47996192, tx: 0 state: 1"
    test_cmd.go:281: (stderr:2) INFO [12-23|01:55:48.979] Trie dumping started                     root=84208a..ae4e13
    test_cmd.go:281: (stderr:2) INFO [12-23|01:55:48.979] Trie dumping complete                    accounts=3 elapsed="48.2µs"
    t8n_test.go:220: args: t8n --output.body "" --output.result stdout --output.alloc stdout --input.alloc ./testdata/3/alloc.json --input.txs ./testdata/3/txs.json --input.env ./testdata/3/env.json --state.fork Berlin
    test_cmd.go:281: (stderr:3) INFO [12-23|01:55:49.059] Trie dumping started                     root=4f19da..1348f4
    test_cmd.go:281: (stderr:3) INFO [12-23|01:55:49.059] Trie dumping complete                    accounts=3 elapsed="80.1µs"
    t8n_test.go:234: test 2: output wrong, have 
.........
```

```
FAIL	github.com/scroll-tech/go-ethereum/cmd/evm	1.125s
--- FAIL: TestConsoleWelcome (0.21s)
--- FAIL: TestAttachWelcome (0.74s)
```

```
INFO [12-23|01:56:04.572] Verification failed error                keyid=284E00B52C269624 error="Invalid signature"
--- FAIL: TestVerification (0.00s)
    --- FAIL: TestVerification/minisig (0.00s)
        version_check_test.go:66: signature could not be verified
```  

```
--- FAIL: TestDecode (0.00s)
    json_test.go:31: input 0x0: got no error, want "hex string of odd length"
    json_test.go:31: input 0x023: got no error, want "hex string of odd length"
``` 

```
--- FAIL: TestColdAccountAccessCost (0.00s)
    runtime_test.go:719: 0: PUSH1 3
    runtime_test.go:719: 1: SELFDESTRUCT 3
    runtime_test.go:721: tescase 6, gas report wrong, step 1, have 3 want 7600
--- FAIL: TestRuntimeJSTracer (0.12s)
    runtime_test.go:894: wrong result for tracer 0 testcase 6, have 
        "1,1,4294964716,4294964716,12"
        want
        "2,2,0,5003,12"
    runtime_test.go:894: wrong result for tracer 1 testcase 6, have 
        "1,1,4294964716,4294964716,0"
        want
        "2,2,0,5003,0"
FAIL
FAIL	github.com/scroll-tech/go-ethereum/core/vm/runtime	0.219s
```

```
--- FAIL: TestTraceBlock (0.01s)
    api_test.go:417: test 1, result mismatch, have
        [{"result":{"gas":21000,"failed":false,"returnValue":"","accountAfter":null,"structLogs":[]}}]
        , want
        [{"result":{"gas":21000,"failed":false,"returnValue":"","structLogs":[]}}]
    api_test.go:417: test 3, result mismatch, have
        [{"result":{"gas":21000,"failed":false,"returnValue":"","accountAfter":null,"structLogs":[]}}]
        , want
        [{"result":{"gas":21000,"failed":false,"returnValue":"","structLogs":[]}}]
    api_test.go:417: test 4, result mismatch, have
        [{"result":{"gas":21000,"failed":false,"returnValue":"","accountAfter":null,"structLogs":[]}}]
        , want
        [{"result":{"gas":21000,"failed":false,"returnValue":"","structLogs":[]}}]
FAIL
FAIL	github.com/scroll-tech/go-ethereum/eth/tracers	0.130s
```

```
--- FAIL: TestCallTracerLegacy (0.00s)
    --- FAIL: TestCallTracerLegacy/selfdestruct (0.03s)
        calltrace_test.go:214: trace mismatch: 
            have &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x10738 Value:0x0 Error:invalid opcode: SELFDESTRUCT Calls:[]}
            want &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x7533 Value:0x0 Error: Calls:[{Type:SELFDESTRUCT From:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE To:0x000000000000000000000000000000000000dEaD Input:0x Output:0x Gas:<nil> GasUsed:<nil> Value:0x4d87094125a369d9bd5 Error: Calls:[]}]}
--- FAIL: TestCallTracerJs (0.00s)
    --- FAIL: TestCallTracerJs/selfdestruct (0.02s)
        calltrace_test.go:214: trace mismatch: 
            have &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x10738 Value:0x0 Error:invalid opcode: SELFDESTRUCT Calls:[]}
            want &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x7533 Value:0x0 Error: Calls:[{Type:SELFDESTRUCT From:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE To:0x000000000000000000000000000000000000dEaD Input:0x Output:0x Gas:0x0 GasUsed:0x0 Value:0x4d87094125a369d9bd5 Error: Calls:[]}]}
--- FAIL: TestCallTracerNative (0.00s)
    --- FAIL: TestCallTracerNative/selfdestruct (0.01s)
        calltrace_test.go:214: trace mismatch: 
            have &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x10738 Value:0x0 Error:invalid opcode: SELFDESTRUCT Calls:[]}
            want &{Type:CALL From:0xb436ba50D378D4BBC8660d312A13dF6af6E89dFB To:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE Input:0x63e4bff40000000000000000000000000024f658a46fbb89d8ac105e98d7ac7cbbaf27c5 Output:0x Gas:0x10738 GasUsed:0x7533 Value:0x0 Error: Calls:[{Type:SELFDESTRUCT From:0x3b873A919aA0512D5A0F09E6dCCEaA4a6727faFE To:0x000000000000000000000000000000000000dEaD Input:0x Output:0x Gas:0x0 GasUsed:0x0 Value:0x4d87094125a369d9bd5 Error: Calls:[]}]}
FAIL
FAIL	github.com/scroll-tech/go-ethereum/eth/tracers/internal/tracetest	0.836s
```

```
--- FAIL: TestWriterCreation (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/deploy.json: no such file or directory
--- FAIL: TestGreeterTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/greeter.json: no such file or directory
--- FAIL: TestTokenTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/token.json: no such file or directory
--- FAIL: TestCallTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/call.json: no such file or directory
--- FAIL: TestCreateTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/create.json: no such file or directory
--- FAIL: TestFailedCallTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/fail_call.json: no such file or directory
--- FAIL: TestDeleteTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/delete.json: no such file or directory
--- FAIL: TestDestructTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/destruct.json: no such file or directory
--- FAIL: TestMutipleTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/multi_txs.json: no such file or directory
```

```
--- FAIL: TestOdrAccountsLes2 (0.09s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
```

```
--- FAIL: TestOdrAccountsLes2 (0.01s)
    odr_test.go:308: checking without ODR
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
```

```
--- FAIL: TestAndroid (14.59s)
    android_test.go:175: gomobile missing, installing it...
    android_test.go:179: initializing gomobile...
    android_test.go:184: initialization took 1.856263889s
    android_test.go:212: gomobile: ANDROID_NDK_HOME specifies /usr/local/lib/android/sdk/ndk/25.1.8937393, which is unusable: unsupported API version 16 (not in 19..33)
    android_test.go:213: failed to run gomobile bind: exit status 1
```

```
--- FAIL: TestParseBytes (0.00s)
    signed_data_internal_test.go:105: test 5: expected !ok, got ok = true with out = 001233
```

**Comment by Lawliet-Chan** [comment]
- github.com/scroll-tech/go-ethereum/cmd/devp2p/internal/ethtest/suite_test.go
  - [ ] TestEthSuite (fail)
   
- github.com/scroll-tech/go-ethereum/cmd/evm/t8n_test.go
  - [x] TestT8n (fail)

- github.com/scroll-tech/go-ethereum/cmd/geth
  - [x] TestConsoleWelcome (fail)
  - [x] TestAttachWelcome (fail)
  - [x] TestVerification (fail)

- github.com/scroll-tech/go-ethereum/common/hexutil
  - [x] TestDecode (fail)

- github.com/scroll-tech/go-ethereum/core/vm/runtime
  - [x] TestColdAccountAccessCost(fail)
  - [x] TestRuntimeJSTracer(fail)

- github.com/scroll-tech/go-ethereum/eth/downloader/downloader_test.go
  - [x] TestFakedSyncProgress66Full (fail)

- github.com/scroll-tech/go-ethereum/eth/tracers/api_test.go
  - [x] TestTraceBlock(fail)

- github.com/scroll-tech/go-ethereum/eth/tracers/internal/tracetest  （Because of `selfdestruct` opcode is invalid, no idea how to fix it) 
  - [x] TestCallTracerLegacy (fail)
  - [x] TestCallTracerJs (fail)
  - [x] TestCallTracerNative (fail)

- github.com/scroll-tech/go-ethereum/internal/utesting   (The follows' errors are all `no such file or directory`, just give these trace json files)
  - [ ] ~~TestWriterCreation~~ 
  - [ ] ~~TestGreeterTx~~ 
  - [ ] ~~TestTokenTx~~ 
  - [ ] ~~TestCallTx~~ 
  - [ ] ~~TestCreateTx~~ 
  - [ ] ~~TestFailedCallTx~~ 
  - [ ] ~~TestDeleteTx~~
  - [ ] ~~TestDestructTx~~
  - [ ] ~~TestMutipleTx~~

- github.com/scroll-tech/go-ethereum/les  （Panic for here https://github.com/scroll-tech/go-ethereum/blob/staging/light/trie.go#L93 )
  - [ ] ~~TestOdrAccountsLes2(panic)~~

- github.com/scroll-tech/go-ethereum/light   (same as above)
  - [ ] ~~TestOdrAccountsLes2(panic)~~

- github.com/scroll-tech/go-ethereum/mobile
  - [x] TestAndroid (fail)

- github.com/scroll-tech/go-ethereum/signer/core
  - [x] TestParseBytes (fail)

File: common/hexutil/hexutil.go [hunk 1]
```diff
@@ -64,10 +64,6 @@ func Decode(input string) ([]byte, error) {
 	if !has0xPrefix(input) {
 		return nil, ErrMissingPrefix
 	}
-	// better compatible with odd size string
```
Comment (Lawliet-Chan) [inline common/hexutil/hexutil.go]: @noel2004  @HAOYUatHZ  Why add these codes before? It causes some tests failed.

**Comment by Lawliet-Chan** [comment]
```
--- FAIL: TestWriterCreation (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/deploy.json: no such file or directory
--- FAIL: TestGreeterTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/greeter.json: no such file or directory
--- FAIL: TestTokenTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/token.json: no such file or directory
--- FAIL: TestCallTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/call.json: no such file or directory
--- FAIL: TestCreateTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/create.json: no such file or directory
--- FAIL: TestFailedCallTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/fail_call.json: no such file or directory
--- FAIL: TestDeleteTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/delete.json: no such file or directory
--- FAIL: TestDestructTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/destruct.json: no such file or directory
--- FAIL: TestMutipleTx (0.00s)
    zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/multi_txs.json: no such file or directory
```
These tests need json trace files, could you please generate them? @noel2004  @HAOYUatHZ

File: common/hexutil/hexutil.go (hunk 1 above)
Comment (noel2004) [reply common/hexutil/hexutil.go]: I guess it require the handled hex strings has '0x' prefix for some legacy reasons? Anyway, enforcing the prefix should has some benefits in developing for avoiding decoding dec or some other type of strings by mistake. So I advice to keep it unless there are too many difficulities for making a workaround.

Comment (Lawliet-Chan) [reply common/hexutil/hexutil.go]: Keep these codes will cause error.  Here it is https://github.com/scroll-tech/go-ethereum/actions/runs/3794039841/jobs/6451737909  
Failed: 
   - signed_data_internal_test.go:105:   TestParseBytes
  
If keep them, must fix the `TestParseBytes`.

**Comment by noel2004** [comment]
>

**Comment by noel2004** [comment]
> ```
> --- FAIL: TestWriterCreation (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/deploy.json: no such file or directory
> --- FAIL: TestGreeterTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/greeter.json: no such file or directory
> --- FAIL: TestTokenTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/token.json: no such file or directory
> --- FAIL: TestCallTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/call.json: no such file or directory
> --- FAIL: TestCreateTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/create.json: no such file or directory
> --- FAIL: TestFailedCallTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/fail_call.json: no such file or directory
> --- FAIL: TestDeleteTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/delete.json: no such file or directory
> --- FAIL: TestDestructTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/destruct.json: no such file or directory
> --- FAIL: TestMutipleTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/multi_txs.json: no such file or directory
> ```
> 
> These tests need json trace files, could you please generate them? @noel2004 @HAOYUatHZ

I prefer do it later: the new zktrie circui would be integrated into common-rs with a new built-in mpt-witness module, which would replace the mpt_witness module in l2geth and vanish the usage of `mpt_witness` field in blocktrace. So we may not need these tests examples anymore.

**Comment by 0xmountaintop** [comment]
> ```
> --- FAIL: TestWriterCreation (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/deploy.json: no such file or directory
> --- FAIL: TestGreeterTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/greeter.json: no such file or directory
> --- FAIL: TestTokenTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/token.json: no such file or directory
> --- FAIL: TestCallTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/call.json: no such file or directory
> --- FAIL: TestCreateTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/create.json: no such file or directory
> --- FAIL: TestFailedCallTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/fail_call.json: no such file or directory
> --- FAIL: TestDeleteTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/delete.json: no such file or directory
> --- FAIL: TestDestructTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/destruct.json: no such file or directory
> --- FAIL: TestMutipleTx (0.00s)
>     zktrie_gen_witness_test.go:28: open blocktraces/mpt_witness/multi_txs.json: no such file or directory
> ```
> 
> These tests need json trace files, could you please generate them? @noel2004 @HAOYUatHZ

actually we do have these traces.

You must forget to init the submodule https://github.com/scroll-tech/go-ethereum/blob/staging/.gitmodules

(But as @noel2004 mention, we need to re-generate them later.)

**Comment by noel2004** [comment]
> * github.com/scroll-tech/go-ethereum/cmd/devp2p/internal/ethtest/suite_test.go
>   
>   * [ ]  TestEthSuite (fail)

Has discovered why `TestEthSuite` fail:

In block 31 the test chain deployed a contract on `0x880EC53Af800b5Cd051531672EF4fc4De233bD5d` for the first time while we have change the codehash calculation from keccak to poseidon and has no any switch to disable it. (So you have a different codehash for this account and of course different state root)

So it may not an easy task to repair it unless we have added the switch to change codehash back to keccak256.

And also I recall that the 'suicide' opeartion would bring the same inconsistent case for we disabled it in our node.

**Comment by 0xmountaintop** [comment]
``

**Pull request closed without merging**